File: py/aoc2024/day3.py
"""
Day 3: Mull It Over
"""

import re

SAMPLE_INPUT_1 = """
xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))
"""
SAMPLE_INPUT_2 = """
xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))
"""

_pattern1 = re.compile(r"mul\((\d+),(\d+)\)")
_pattern2 = re.compile(r"(?:\A|do\(\))(.*?)(?:\Z|don't\(\))", re.S)


def part1(data: str) -> int:
    """
    >>> part1(SAMPLE_INPUT_1)
    161
    """
    return sum(int(m.group(1)) * int(m.group(2)) for m in _pattern1.finditer(data))


def part2(data: str) -> int:
    """
    >>> part2(SAMPLE_INPUT_2)
    48
    """
    return sum(part1(m.group(1)) for m in _pattern2.finditer(data))


parts = (part1, part2)

File: rs/src/day3.rs
use std::iter::successors;

pub fn part1(data: &str) -> i32 {
    data.split("mul(")
        .skip(1)
        .filter_map(|data| {
            let (a, data) = data.split_once(',')?;
            let a = a.parse::<i32>().ok()?;
            let (b, _) = data.split_once(')')?;
            let b = b.parse::<i32>().ok()?;
            Some(a * b)
        })
        .sum()
}

pub fn part2(data: &str) -> i32 {
    fn split_end(data: &str) -> (&str, &str) {
        data.split_once("don't()").unwrap_or((data, ""))
    }
    successors(Some(split_end(data)), |(_, data)| {
        Some(split_end(data.split_once("do()")?.1))
    })
    .map(|(data, _)| part1(data))
    .sum()
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use pretty_assertions::assert_eq;

    static EXAMPLE1: &str = indoc! {"
        xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))
    "};
    static EXAMPLE2: &str = indoc! {"
        xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))
    "};

    #[test]
    fn part1_examples() {
        assert_eq!(161, part1(EXAMPLE1));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(48, part2(EXAMPLE2));
    }
}
